### PR TITLE
pass operator data into Structure::MultigridPreconditioner update/init

### PR DESCRIPTION
**Pull request by richardschu**
The scaling factor for the mass operator was not passed into the MG levels in the structure module. At init and update, the scaling factors (and the evaluation time for the operators) are now passed to all levels and kept up to date.
For the example applications, this barely made a difference ... but of course that depends on the parameters.

### CODE REVIEW

File: include/exadg/structure/driver.cpp
Comment (richardschu) [inline include/exadg/structure/driver.cpp:115]: here i needed to get the scaling factor from thetime integrator. hence I made it public, see `time_int_gen_alpha_base.h`

File: include/exadg/structure/preconditioners/multigrid_preconditioner.cpp
Comment (richardschu) [inline include/exadg/structure/preconditioners/multigrid_preconditioner.cpp:238]: scaling factors are correct at first init with this

File: include/exadg/structure/spatial_discretization/operator.cpp
Comment (richardschu) [inline include/exadg/structure/spatial_discretization/operator.cpp:722]: we set the factor to zero, so store the value we had before and reset it after computing the initial acceleration

File: include/exadg/structure/spatial_discretization/operator.cpp
Comment (richardschu) [inline include/exadg/structure/spatial_discretization/operator.cpp:933]: note that we update the preconditioner before the actual solve!

File: include/exadg/structure/spatial_discretization/operators/linear_operator.cpp
Comment (richardschu) [inline include/exadg/structure/spatial_discretization/operators/linear_operator.cpp:53]: i can remove these, I had some trusty std::cout here

File: include/exadg/structure/spatial_discretization/operators/nonlinear_operator.cpp
Comment (richardschu) [inline include/exadg/structure/spatial_discretization/operators/nonlinear_operator.cpp:295]: same here

File: include/exadg/structure/time_integration/time_int_gen_alpha.cpp
Comment (richardschu) [inline include/exadg/structure/time_integration/time_int_gen_alpha.cpp:152]: preconditioner might be updated for linear (instat) solve as well -> scaling for adaptive timestepping and large dt variations?

File: include/exadg/time_integration/time_int_gen_alpha_base.h
Comment (richardschu) [inline include/exadg/time_integration/time_int_gen_alpha_base.h:55]: made this public to call it in the structure driver

**Review by richardschu — COMMENTED** [review]
what do you think @nfehn ?

File: include/exadg/structure/preconditioners/multigrid_preconditioner.cpp
Comment (nfehn) [inline include/exadg/structure/preconditioners/multigrid_preconditioner.cpp:154]: Did you figure out the precise reason for `set_time()`? This would be important because if time does not have an impact on the linear system of equations, we should avoid updating the preconditioner (for reasons of computational costs).

File: include/exadg/structure/preconditioners/multigrid_preconditioner.cpp
Comment (nfehn) [reply include/exadg/structure/preconditioners/multigrid_preconditioner.cpp:73]: I have just seen that we do it exactly the same way in the momentum operator for incompressible flows. So should be fine as is. I think the responsibility is here in the hands of the user, who has to specify how often to update the preconditioner in the application file. If no update is needed (for certain model problems), the user has to deactivate this in the application to achieve optimal performance.

File: include/exadg/structure/spatial_discretization/operator.cpp
Comment (nfehn) [inline include/exadg/structure/spatial_discretization/operator.cpp:442]: I assume this is in analogy to how it is done in other modules such as IncNS? I think there is no need to compute an active scaling factor, because it will never be used for a steady problem (because the operator itself contains an if-statement `if(unsteady)`)? If so, I suggest to skip the first two lines.

BTW: The circumstance that a setup of the operator is done in `setup_solvers()` is exactly the Henne-Ei-Problem we have discussed in person.

File: include/exadg/structure/spatial_discretization/operator.cpp
Comment (nfehn) [inline include/exadg/structure/spatial_discretization/operator.cpp:772]: I would write each of these lines directly after evaluating the elasticity operator in the respective if/else-branches above so that it is more clear for which parts of the code this is relevant.

File: include/exadg/structure/spatial_discretization/operator.cpp
Comment (nfehn) [inline include/exadg/structure/spatial_discretization/operator.cpp:721]: No need to define this outside. We can simply write `double const` inside the if-branch and the else-branch. Makes the code easier to read in my opinion (and in the sense of @btemuer, the use of `scaling_factor_mass` is here again in a `const` way).

File: include/exadg/structure/spatial_discretization/operators/linear_operator.cpp
Comment (nfehn) [reply include/exadg/structure/spatial_discretization/operators/linear_operator.cpp:53]: I actually like it to place brackets here (should also be in line with our coding conventions in the wiki).

File: include/exadg/structure/time_integration/driver_steady_problems.cpp
Comment (nfehn) [inline include/exadg/structure/time_integration/driver_steady_problems.cpp:128]: We have to rethink this critically. In my opinion, this means we initialize the multigrid preconditioner (or, better, the smoothers, coarse grid solver) twice or a steady linear problem: once in the setup-routines, and once within the solver call. However, this is unnecessary work. So we should check if this is the case, and if yes, pass `false` to this function here with a comment that the preconditioner has already been initialized.

Note that this is different for the Newton solver above, because there the preconditioner can be updated every n-th Newton iteration.

File: include/exadg/structure/preconditioners/multigrid_preconditioner.cpp
Comment (richardschu) [reply include/exadg/structure/preconditioners/multigrid_preconditioner.cpp:73]: This way, the steady linear solver does not do the update - as it would otherwise do the update twice (in a row, basically)!

Comment (richardschu) [reply include/exadg/structure/preconditioners/multigrid_preconditioner.cpp:73]: The cases where we do additional inits are now:
Steady nonlinear solver (+1 initially)
Unsteady nonlinear solver (+ 1 initially)
QuasiStatic(+1 initially)
One can also not skip the update in the first (quasi-)time step, since the update might be triggered every *n* Newton iterations or once the solver has converged.
One solution would be to add another flag to the `update` in the Newton solver, which checks if we are in the very first step after init! What do you think? ... changes some other files as well though ... then we could also think about checking the case `update_once_newton_converged` and the other flags (doing two head-to-head updates basically, with updated iterate/solution vector potentially?). Sounds more like a separate PR to me ... updating that Newton flags and (quasi-)time steppers?

File: include/exadg/structure/preconditioners/multigrid_preconditioner.cpp
Comment (richardschu) [reply include/exadg/structure/preconditioners/multigrid_preconditioner.cpp:154]: So fine? This is only triggered if `unsteady` evaluates to true ... and the time is a double ... so probably thats not the problem is it? the time is never used to trigger a preconditioner update? and updating the time variable is not costly, is it?
Anyways I have not found out what the `time` changes. I can think of changing a stiffness coefficient over time (initial damping and initially stiffer springs help a ton in cardiovasc scenarios!). For now, I do not see a reason this is needed -  maybe it is motivated by someone expecting `update` to update the time of the operators as well, which in itself sounds reasonable to me :)

File: include/exadg/structure/time_integration/driver_steady_problems.cpp
Comment (richardschu) [reply include/exadg/structure/time_integration/driver_steady_problems.cpp]: well actually this evaluates to false here (steady, linear case). so I just put false. `param.update_preconditioner` seemed like here we have some control over that which we do not ...

File: include/exadg/structure/time_integration/time_int_gen_alpha.cpp
Comment (richardschu) [reply include/exadg/structure/time_integration/time_int_gen_alpha.cpp]: I added (this->time_step_number > 1) so it is exactly what you would expect. User has full control via the timestep control of prec update and the initial update directly after init is skipped ---> no additional updates!

**Review by richardschu — COMMENTED** [review]
The update is only triggered for unsteady or nonlinear cases (as before), skipping now one initial update for linear solver in the time dep case. For the nonlinear solver this is not possible, since we might want some updates in the first solution step (every n Newton iterations or once Newton is converged).

File: include/exadg/structure/time_integration/driver_steady_problems.cpp
Comment (richardschu) [inline include/exadg/structure/time_integration/driver_steady_problems.cpp:124]: Note that in the nonlinear case above, we have to pass `param.update_preconditioner` to even allow updates, as the Newton solver checks with an && every *n* iterations or at the end, if the preconditioner should be updated. We could do a follow-up PR to get rid of all the head-to-head calls of the preconditioner update...
For my target applications (nonlinear time-dependent problems + FSI), the effect is probably not sooo big.

File: include/exadg/structure/preconditioners/multigrid_preconditioner.cpp
Comment (nfehn) [reply include/exadg/structure/preconditioners/multigrid_preconditioner.cpp:73]: Currently, I think we should not let the Newton solver take the responsibility for this. I think this aspect is a convention between the spatial operator (initialization of solvers/preconditioners) and the time integrator (where we have the modulo operator, including a convention regarding the first time step for which we want to initiate an update).

File: include/exadg/structure/preconditioners/multigrid_preconditioner.cpp
Comment (nfehn) [reply include/exadg/structure/preconditioners/multigrid_preconditioner.cpp:154]: Then, I suggest to keep it to be prepared for the future / prevent foreseeable bugs.

File: include/exadg/structure/spatial_discretization/operator.cpp
Comment (nfehn) [reply include/exadg/structure/spatial_discretization/operator.cpp]: NB means "nota bene" (something like "Please note that ...")

File: include/exadg/structure/spatial_discretization/operator.cpp
Comment (nfehn) [inline include/exadg/structure/spatial_discretization/operator.cpp:740]: I think one could even shift this line even before evaluation the body force operator. This would make the "manipulation" more local in the code.

File: include/exadg/structure/spatial_discretization/operator.cpp
Comment (nfehn) [inline include/exadg/structure/spatial_discretization/operator.cpp:770]: same as here https://github.com/exadg/exadg/pull/404/files#r1175037799

File: include/exadg/structure/time_integration/driver_steady_problems.cpp
Comment (nfehn) [reply include/exadg/structure/time_integration/driver_steady_problems.cpp:128]: Could you write `false /* update of preconditioner */` to be line with the previous two arguments? In case we do not pass variables with a concrete name but "numbers", it helps to understand the code or interpret the parameters passed to a function (no need to open the declaration of that function).

File: include/exadg/structure/time_integration/time_int_gen_alpha.cpp
Comment (nfehn) [reply include/exadg/structure/time_integration/time_int_gen_alpha.cpp:179]: I would skip `&& (this->time_step_number > 1)` here and do it identically to the nonlinear case. My argument is that the number of updates over a whole unsteady simulation is `N_dt / N_update_interval >> 1`, so the increase in costs is minimal. (Note that the currrent code is potentially problematic because there might come other reasons for preconditioner updates in the future (apart from scaling_factor_mass). Strictly avoiding the preconditioner update in the first call could then cause robustness issues, a bug that might be hard to find once we have forgotton that we have added  `&& (this->time_step_number > 1)` in the present PR.)

File: include/exadg/grid/grid_motion_elasticity.h
Comment (nfehn) [inline include/exadg/grid/grid_motion_elasticity.h:98]: I am not sure if this is what we want to have (of course, it would then also be a bug in the nonlinear case above): I think we want to trigger the update of the preconditioner for the ALE mesh motion solver only every `N_update_interval`th time step (using a logic similar as in the single-field time integration schemes). This logic is not implemented inside `pde_operator->solve_(non)linear()`, right? Does this imply that we update the preconditioner for every ALE mesh motion solve (in every FSI iteration) if `param.update_preconditioner == true`? This could become very expensive.

File: include/exadg/structure/time_integration/time_int_gen_alpha.cpp
Comment (nfehn) [reply include/exadg/structure/time_integration/time_int_gen_alpha.cpp]: see here: https://github.com/exadg/exadg/pull/404#discussion_r1175059454

Comment (richardschu) [reply include/exadg/structure/time_integration/time_int_gen_alpha.cpp]: Perfectly fine with that, as this was my initial thought :+1: ... that one saved update is probably not making the kraut fett!

File: include/exadg/structure/time_integration/driver_steady_problems.cpp
Comment (richardschu) [reply include/exadg/structure/time_integration/driver_steady_problems.cpp:128]: I added the comment!

File: include/exadg/structure/spatial_discretization/operator.cpp
Comment (richardschu) [reply include/exadg/structure/spatial_discretization/operator.cpp:740]: Ok, now it is as close as possible - even though it fit on one screen before ; )

File: include/exadg/grid/grid_motion_elasticity.h
Comment (richardschu) [reply include/exadg/grid/grid_motion_elasticity.h:98]: the grid mapping is constructed from the initial configuration (Lagrangian frame) right?
There are several aspects to this:
1) is the "mesh stiffness" updated depending on the actual displacement or are the parameters computed once initially and kept constant?
2) time step size might vary
3) nonlinear mesh motion model might be used

`pde_opereator->solve_(non)linear()` updates if the passed flag is `true` and in the nonlinear case additionally at Newton convergence or every _n_th Newton step.
As the mapping is a critical ingredient in practical applications, I would actually leave the choice to the user and allow setting the latter two flags in the nonlinear case + a time step logic like you mentioned. So I would  recommend using the same logic as in `time_int_gen_alpha.cpp`, line 149:

`  bool const update_preconditioner =
    this->param.update_preconditioner &&
    ((this->time_step_number - 1) % this->param.update_preconditioner_every_time_steps == 0);`

`pde_operator->solve_nonlinear()` takes care of the rest :100: 

Agree?

Comment (richardschu) [reply include/exadg/grid/grid_motion_elasticity.h:98]: Thinking about this again, I realized that the current grid motion techniques are (i) all solving stationary problems, (ii) the Lagrangian reference frame is considered as base and (iii) the parameters for "grid stiffening" do not depend on the solution itself, so for poisson and linear elasticity operaetors, we should deactivate the preconditioner update completely, shouldn't we? For the nonlinear elasticity case it makes sense though ...
Can you confirm (i)-(iii) @nfehn ?

File: include/exadg/convection_diffusion/time_integration/time_int_bdf.cpp
Comment (nfehn) [inline include/exadg/convection_diffusion/time_integration/time_int_bdf.cpp:72]: I have problems understanding this comment, probably the notion "ALE update" is not fully clear to me at this point.

File: include/exadg/grid/grid_motion_elasticity.h
Comment (richardschu) [inline include/exadg/grid/grid_motion_elasticity.h:79]: the update is only done once per time step (if the update is desired every _n_ time steps).

File: include/exadg/grid/grid_motion_elasticity.h
Comment (richardschu) [inline include/exadg/grid/grid_motion_elasticity.h:121]: linear elasticity leads to a constant matrix, so do not update.

File: include/exadg/grid/grid_motion_interface.h
Comment (richardschu) [inline include/exadg/grid/grid_motion_interface.h:55]: interface is now done with time step number

File: include/exadg/incompressible_flow_with_transport/driver.cpp
Comment (richardschu) [inline include/exadg/incompressible_flow_with_transport/driver.cpp:510]: similar to the other drivers, we give the actual time step number (in case one uses a `GridMotionElasticity` replacing the currently used `GridMotionFunction` or `GridMotionPoisson`), see `Driver::setup()`. For the currently possible `GridMotion` objects, this would not result in an update. Alternatively we would pass `dealii::numbers::invalid_unsigned_int` or simply 0 in the drivers for iNS and ConvDiff.

File: include/exadg/incompressible_navier_stokes/spatial_discretization/spatial_operator_base.cpp
Comment (richardschu) [inline include/exadg/incompressible_navier_stokes/spatial_discretization/spatial_operator_base.cpp:1472]: The `move_grid()` and `move_grid_and_update_dependent_data_structures()` are used during initialization/postprocessing as far as I can tell?
`grep -rnw -e 'move_grid'` in `exadg/include` gives
"
exadg/convection_diffusion/spatial_discretization/interface.h:108:  move_grid(double const & time) const = 0;
exadg/convection_diffusion/spatial_discretization/operator.h:202:  move_grid(double const & time) const;
exadg/convection_diffusion/spatial_discretization/operator.cpp:821:Operator<dim, Number>::move_grid(double const & time) const
exadg/convection_diffusion/time_integration/time_int_bdf.cpp:72:    pde_operator->move_grid(this->get_time());
exadg/convection_diffusion/time_integration/time_int_bdf.cpp:80:        pde_operator->move_grid(this->get_previous_time(i));
exadg/convection_diffusion/time_integration/time_int_bdf.cpp:141:    pde_operator->move_grid(this->get_time());
exadg/convection_diffusion/time_integration/time_int_bdf.cpp:154:      pde_operator->move_grid(this->get_previous_time(i));
exadg/incompressible_navier_stokes/spatial_discretization/spatial_operator_base.cpp:1466:SpatialOperatorBase<dim, Number>::move_grid(double const & time) const
exadg/incompressible_navier_stokes/spatial_discretization/spatial_operator_base.h:389:  move_grid(double const & time) const;
exadg/incompressible_navier_stokes/time_integration/time_int_bdf_coupled_solver.cpp:67:    pde_operator->move_grid(this->get_time());
exadg/incompressible_navier_stokes/time_integration/time_int_bdf_coupled_solver.cpp:82:      pde_operator->move_grid(this->get_previous_time(i));
exadg/incompressible_navier_stokes/time_integration/time_int_bdf_pressure_correction.cpp:147:    pde_operator->move_grid(this->get_time());
exadg/incompressible_navier_stokes/time_integration/time_int_bdf_pressure_correction.cpp:160:      pde_operator->move_grid(this->get_previous_time(i));
exadg/incompressible_navier_stokes/time_integration/time_int_bdf.cpp:99:    operator_base->move_grid(this->get_time());
exadg/incompressible_navier_stokes/time_integration/time_int_bdf.cpp:107:        operator_base->move_grid(this->get_previous_time(i));
exadg/incompressible_navier_stokes/time_integration/time_int_bdf_dual_splitting.cpp:139:    pde_operator->move_grid(this->get_time());
exadg/incompressible_navier_stokes/time_integration/time_int_bdf_dual_splitting.cpp:152:      pde_operator->move_grid(this->get_previous_time(i));

"

So I would rather stick to the convention of triggering the update only in `Driver::ale_update()`.

File: include/exadg/grid/grid_motion_function.h
Comment (richardschu) [inline include/exadg/grid/grid_motion_function.h:51]: you prefer 0 whenever I put `dealii::numbers::invalid_unsigned int`? it is `static_cast<unsigned int>(-1)`, this way one would now immediately get that the `time_step_number` is fishy when debugging (and having an enormous unsigned int unexpectedly)!

File: include/exadg/convection_diffusion/spatial_discretization/operator.cpp
Comment (nfehn) [inline include/exadg/convection_diffusion/spatial_discretization/operator.cpp:827]: The problem here is that the function `move_grid()` is public and can be used in a different context that our current `Driver`. In this sense, `Operator` is a low level module, whose implementation should not depend on our current high-level implementation in `Driver`. As a consequence, there should actually also not appear a comment on `Driver` in this class (in an ideal world).

I think the origin of this problem is that the two functions `move_grid()` and `move_grid_and_update_dependent_data_structures()` are a sub-optimal design. I think we should remove `move_grid()` and let the time integrator directly operate on `grid_motion`. The function `move_grid_and_update_dependent_data_structures()` needs to be changed to `update_spatial_discretization_after_grid_motion()` (i.e. the first part `grid_motion->update()` should no longer be part of that function). This way, there is only one possibility to move the grid in the future.

In order to proceed with the present PR, I suggest that I could address this in the future once the present PR is merged.

File: include/exadg/incompressible_navier_stokes/spatial_discretization/spatial_operator_base.cpp
Comment (nfehn) [reply include/exadg/incompressible_navier_stokes/spatial_discretization/spatial_operator_base.cpp:1472]: see comment https://github.com/exadg/exadg/pull/404/files#r1177448382

**Review by nfehn — APPROVED** [review]

File: include/exadg/convection_diffusion/spatial_discretization/operator.cpp
Comment (nfehn) [reply include/exadg/convection_diffusion/spatial_discretization/operator.cpp:827]: I created issue #412 and will therefore resolve the conversation here

**Comment by nfehn** [comment]
I will merge once the checks have passed

**Pull request merged**
